### PR TITLE
Add ability to ignore tests depending on spark shim version

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Adds the ability to ignore scala tests conditionally, depending on the version of Spark we are testing with.

Also closes https://github.com/NVIDIA/spark-rapids/issues/382

### CODE REVIEW

**Comment by tgravescs** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/StringFallbackSuite.scala [hunk 1]
```diff
@@ -29,7 +29,17 @@ class StringFallbackSuite extends SparkQueryCompareTestSuite {
     "RegExpReplace",
     nullableStringsFromCsv, execsAllowedNonGpu = Seq("ProjectExec", "Alias",
       "RegExpReplace", "AttributeReference", "Literal")) {
-    frame => frame.selectExpr("regexp_replace(strings,null,'D')")
+    frame => {
+      ShimLoader.getSparkShims.getSparkShimVersion match {
+        case SparkShimVersion(major, minor, _) =>
+          // this test is not valid in Spark 3.1 and later because the expression is
+          // NullIntolerant and gets replaced with a null literal instead
+          val isValidTestForSparkVersion = major <= 3 && minor == 0
```
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/StringFallbackSuite.scala:37]: Although it seems unnecessary to have this separate variable, it makes the output message more obvious:

```
Test Canceled: isValidTestForSparkVersion was false
```

**Comment by andygrove** [comment]
I wasn't sure how this would work with Databricks or what the version numbers would be. The case classes might give us more flexibility than strings and could perhaps use these instead of the string for version name?

I pushed another commit to demonstrate how I imagined this working for databricks.

**Comment by tgravescs** [comment]
we can use the case classes as the real version and then have it have a function toString, that is used in SparkShimServiceProvider for now.  That way we don't have the version in 2 places.

I would like to add it to the databricks version as well. I don't want to have to fix it afterwards. Have it run the tests for it as it is now and the test is passing. They only changed this in 3.1 and so far 7.0 databricks runtime doesn't have that change.  If the test ends up failing we can handle it separate but I would at least like to make an attempt here otherwise compilation will fail.  The databricks version is just 3.0.0-databricks. It is not 7.0 which is the runtime version. 

The downside to having different case classes is demonstrated here, where you would have to match on both. If we had a "vendor" field in SparkShimVersion you could conditionalize that. In this case your check (noticed extra _) 
SparkShimVersion(major, minor, _, _) => major == 3 && minor == 0
would still work and cover databricks.
I can see reasons to do it both ways though. For instance this way would allow the Databricks version to have a separate runtime field in it as well.  I'm fine with leaving it like this for now with different case classes, but please add the version to the 300db Shim layer as DatabricksShimVersion(3,0,0) and have the test run for it.

I think either way we will have to enhance the databricks version and probably this anyway to handle cases where the version is like 3.0.0.0 - but there is a separate jira for that.

**Comment by andygrove** [comment]
Thanks @tgravescs I hadn't realized that the databricks shim was already merged in. I have made those updates although this change has turned out larger than I originally thought. Let me know what you think. I'm happy to take a different approach if this is too disruptive.

File: shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/SparkShimServiceProvider.scala [hunk 2]
```diff
@@ -16,14 +16,16 @@
 
 package com.nvidia.spark.rapids.shims.spark300
 
-import com.nvidia.spark.rapids.SparkShims
+import com.nvidia.spark.rapids.{SparkShims, SparkShimVersion}
 
-class SparkShimServiceProvider extends com.nvidia.spark.rapids.SparkShimServiceProvider {
+object SparkShimServiceProvider {
+  val VERSION = SparkShimVersion(3,0,0)
```
Comment (tgravescs) [inline shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/SparkShimServiceProvider.scala:22]: nit add spaces after ,'s

File: shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/SparkShimServiceProvider.scala [hunk 3]
```diff
@@ -16,14 +16,16 @@
 
 package com.nvidia.spark.rapids.shims.spark301
 
-import com.nvidia.spark.rapids.SparkShims
+import com.nvidia.spark.rapids.{SparkShims, SparkShimVersion}
 
+object SparkShimServiceProvider {
+  val VERSION = SparkShimVersion(3,0,1)
```
Comment (tgravescs) [inline shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/SparkShimServiceProvider.scala:22]: nit add spaces after ,

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/SparkShimServiceProvider.scala [hunk 4]
```diff
@@ -16,14 +16,17 @@
 
 package com.nvidia.spark.rapids.shims.spark310
 
-import com.nvidia.spark.rapids.SparkShims
+import com.nvidia.spark.rapids.{SparkShims, SparkShimVersion}
 
-class SparkShimServiceProvider extends com.nvidia.spark.rapids.SparkShimServiceProvider {
+object SparkShimServiceProvider {
+  val VERSION = SparkShimVersion(3,1,0)
```
Comment (tgravescs) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/SparkShimServiceProvider.scala:22]: nit ad spaces after ,

**Comment by tgravescs** [comment]
overall I think this is fine, few minor nits. We can enhance it as we need.

**Comment by tgravescs** [comment]
build

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
@tgravescs I addressed those nits

**Pull request merged**